### PR TITLE
Generalize mean to other types

### LINKED ISSUE #2471: math.mean() for openArray[int]
Ready for release of Nim v1.0,
should the Math module be extended to allow for a wider use of types, eg the mean of a seq or array or ints, rather than only handling floats?

### PR DESCRIPTION
**Pull request by jsanjuas**
Should fix #2471

### CODE REVIEW

**Comment by ReneSac** [comment]
Note that this proc will give an overflow error during calculation for most big integer arrays/seqs. The ideal solution here would be to use a BigInt to accumulate the sum result... but there is no BigInt in the stdlib (yet?).

**Comment by Araq** [comment]
Looks good to me but now variance() needs to do the same...

**Comment by ReneSac** [comment]
I would suggest to at least document those possible shortcomings. 

An alternative to bignums is to make the sum as float. You will just have the normal float sum minor precision issues that the current sum() doesn't deal anyway, but it would be much slower due to the float conversion for each element. But it at least it won't overflow

**Comment by Araq** [comment]
@ReneSac That's what this PR does, it converts ``T`` to a float.

**Comment by ReneSac** [comment]
He calls toFloat only after the sum. And that template looks strange. It would be good to have some tests included.

**Comment by jsanjuas** [comment]
Yes, I agree that converting to float prior to summing is a more sane approach. I can't comment on how strange the template is or isn't, but it does look clean to me. I'll revisit this PR over the weekend.

File: lib/pure/math.nim [hunk 1]
```diff
@@ -114,18 +114,23 @@ proc sum*[T](x: openArray[T]): T {.noSideEffect.} =
   ## If `x` is empty, 0 is returned.
   for i in items(x): result = result + i
 
-proc mean*(x: openArray[float]): float {.noSideEffect.} = 
-  ## computes the mean of the elements in `x`. 
+template toFloat(f: float): float = f
```
Comment (flaviut) [inline lib/pure/math.nim]: This is redundant.

File: lib/pure/math.nim [hunk 2]
```diff
@@ -114,18 +114,23 @@ proc sum*[T](x: openArray[T]): T {.noSideEffect.} =
...
+
+proc mean*[T](x: openArray[T]): float {.noSideEffect.} =
+  ## computes the mean of the elements in `x`, which are first converted to floats.
   ## If `x` is empty, NaN is returned.
-  result = sum(x) / toFloat(len(x))
+  ## ``toFloat(x: T): float`` must be defined.
+  for i in items(x): result = result + toFloat(i)
```
Comment (flaviut) [inline lib/pure/math.nim]: for i in items(x): result = result + float(i)

is better

Comment (flaviut) [inline lib/pure/math.nim]: actually, never mind. I see what's going on here.

File: lib/pure/math.nim [hunk 3]
```diff
@@ -114,18 +114,23 @@ proc sum*[T](x: openArray[T]): T {.noSideEffect.} =
...
-proc variance*(x: openArray[float]): float {.noSideEffect.} = 
+proc variance*[T](x: openArray[T]): float {.noSideEffect.} =
   ## computes the variance of the elements in `x`. 
   ## If `x` is empty, NaN is returned.
+  ## ``toFloat(x: T): float`` must be defined.
   result = 0.0
   var m = mean(x)
-  for i in 0 .. high(x):
-    var diff = x[i] - m
+  for i in items(x):
```
Comment (flaviut) [inline lib/pure/math.nim]: Although this might be better as `for i in x:`.

**Pull request merged**